tiledb/sm/query/reader.cc

```diff
@@ -1054,97 +1054,159 @@ Status Reader::copy_var_cells(
   auto buffer_var = (unsigned char*)it->second.buffer_var_;
   auto buffer_size = it->second.buffer_size_;
   auto buffer_var_size = it->second.buffer_var_size_;
-  uint64_t buffer_offset = 0;
-  uint64_t buffer_var_offset = 0;
   uint64_t offset_size = constants::cell_var_offset_size;
-  uint64_t cell_var_size;
   auto type = array_schema_->type(attribute);
   auto fill_size = datatype_size(type);
   auto fill_value = constants::fill_value(type);
   assert(fill_value != nullptr);
 
-  // Copy cells
-  for (const auto& cr : cell_ranges) {
-    auto cell_num_in_range = cr.end_ - cr.start_ + 1;
-    // Check if offset buffers can fit the result
-    if (buffer_offset + cell_num_in_range * offset_size > *buffer_size) {
-      read_state_.overflowed_ = true;
-      return Status::Ok();
-    }
+  // Compute the destinations of offsets and var-len data in the buffers.
+  std::vector<std::vector<uint64_t>> offset_offsets_per_cr;
+  std::vector<std::vector<uint64_t>> var_offsets_per_cr;
+  uint64_t total_offset_size, total_var_size;
+  RETURN_NOT_OK(compute_var_cell_destinations(
+      attribute,
+      cell_ranges,
+      &offset_offsets_per_cr,
+      &var_offsets_per_cr,
+      &total_offset_size,
+      &total_var_size));
+
+  // Check for overflow and return early (without copying) in that case.
+  if (total_offset_size > *buffer_size || total_var_size > *buffer_var_size) {
+    read_state_.overflowed_ = true;
+    return Status::Ok();
+  }
 
-    // Handle empty range
-    if (cr.tile_ == nullptr) {
-      // Check if result can fit in the buffer
-      if (buffer_var_offset + cell_num_in_range * fill_size >
-          *buffer_var_size) {
-        read_state_.overflowed_ = true;
-        return Status::Ok();
-      }
+  // Copy cell ranges in parallel.
+  const auto num_cr = cell_ranges.size();
+  auto statuses = parallel_for(0, num_cr, [&](uint64_t cr_idx) {
+    const auto& cr = cell_ranges[cr_idx];
+    const auto& offset_offsets = offset_offsets_per_cr[cr_idx];
+    const auto& var_offsets = var_offsets_per_cr[cr_idx];
+
+    // Get tile information, if the range is nonempty.
+    uint64_t* tile_offsets = nullptr;
+    unsigned char* tile_var_data = nullptr;
+    uint64_t tile_cell_num = 0;
+    uint64_t tile_var_size = 0;
+    if (cr.tile_ != nullptr) {
+      const auto& tile_pair = cr.tile_->attr_tiles_.find(attribute)->second;
```

Comment:
Should this go outside the `parallel_for` call?

Reply:
I don't think we can, can we? It depends on `cr`.